graphprotocol/contracts
# feat: e2e protocol deployment testing

### Motivation

Currently after deploying the protocol, configuring it and starting it we don't have an automated way of testing if everything went smoothly. Our test suite consists of individually testing the smart contracts, using mock data and fixtures and a completely different "deployment" procedure. 

This PR introduces "e2e" tests. These tests are ran against a fully deployed and configured version of the protocol running on a previously deployed local testnet. It is a WIP, but some ideas related to protocol deployment are:

- Deployment
   - Contracts deployed with the correct parameters (as configured in graph config file)
   - Protocol roles correctly configured (governor, arbiter, subgraphoracle, etc) 
   - Contract ownership of governed contracts (i.e: deployer doesn't own them)
 
Additionally, we can incorporate tests to cover the "traditional actions" such as curating, creating subgraphs, allocations, etc (and expand with L2 later).

### Usage

Run with: `yarn test:e2e`. 
The script will invoke several hardhat tasks, including:
- Start a hardhat node (`localhost`)
- Run `migrate:accounts` hardhat task to create keys for all protocol roles (deployer, governor, arbiter, etc). This currently doesn't support multisig accounts.
- Run `migrate` hardhat task to deploy the protocol
- Run `migrate:ownership` hardhat task to transfer ownership of governed contracts to the governor
- Run `migrate:unpause` to unpause the protocol
- Run e2e tests 

### Changes

- Add post deployment transactions to graph config file (see `graph.localhost.yml`)
   - Deployer transfer ownership of contracts (GraphToken, Controller, SubgraphNFT, GraphProxyAdmin)
   - Deployer renounce minter
   - Set Pause guardian on Controller
   - Set subgraph availability oracle on Rewards manager
- Add new entities to graph config file
   - availabilityOracle: address of the Subgraph Availability Oracle
   - pauseGuardian: address of the protocol pause guardian
   - edgeAndNode: address of the AllocationExchange owner 
- Tests
    - Deployment
        - Test contract configuration matches graph config description
        - Test contract ownership of governed contracts
        - Test basic roles configuration. Ensure the protocol is configured to allow their actions (pause guardian, authority, availability oracle, arbitrator)
        - Test protocol is running (unpaused)
- Hardhat tasks
   - Add task to create protocol accounts and store their addresses in a graph config file (only for localhost/e2e)
   - Add task to accept ownership of protocol contracts on behalf of the governor (only for localhost/e2e)
   - Add task to unpause the protocol


Signed-off-by: Tomás Migone <tomas@edgeandnode.com>
